addons/necropost-highlighter/userscript.js

```diff
@@ -0,0 +1,294 @@
+/**
+ * @author RT_Borg
+ * 
+ * Necropost Highlighter
+ * 
+ * Uses topic id as a proxy for age, which doesn't require loading other pages
+ * to read actual dates.
+ *
+ * The number of new topics per month on Scratch forums is fairly stable over time.
+ * Since the topic interval doesn't need to be precise to determine a likely necropost,
+ * this ends up working very well.
+ *
+ * Inspects topic ids of all posts on the page and highlights any topics significantly
+ * older than surrounding posts.
+ * 
+ * Works with:
+ *   - Addon: Website dark mode and customizable colors
+ *   - Addon: Infinite scrolling
+ *   - Standard and Mobile websites
+ */
+
+export default async function ({ addon, global, console }) {
+	// Default is a little over a month's worth of topics. Reasonably stable through 2022.
+	// There's a lot of tolerance. Half or twice as many topics filter pretty much the same.
+	const defaultTopicInterval = 7500;
+	const defaultForums = [
+		"New Scratchers",
+		"Help with Scripts",
+		"Questions about Scratch",
+		"Bugs and Glitches",
+		"Project Ideas"
+	];
+	// Forums beyond these seem very poor options,
+	// as many posts live long or are appropriately revived
+	const customForumSettingToForum = {
+		"applyToNewScratchers": "New Scratchers",
+		"applyToHelpWithScripts": "Help with Scripts",
+		"applyToQuestionsAboutScratch": "Questions about Scratch",
+		"applyToBugsAndGlitches": "Bugs and Glitches",
+		"applyToProjectIdeas": "Project Ideas",
+		"applyToRequests": "Requests",
+		"applyToSuggestions": "Suggestions",
+		"applyToOpenSourceProjects": "Open Source Projects"
+	};
+	const searchResultsPageName = "Search Results"; // recent posts and unanswered posts
+	let isMobileSite = await determineIsMobileSite();
+	let forumOrSearchPageName = await determineForumOrSearchPageName();
+
+	highlightNecropostsIfOnSelectedForumPage();
+	
+	addon.settings.addEventListener("change", onAddonSettingsChange);
+	addon.self.addEventListener("disabled", onDisabled);
+	addon.self.addEventListener("reenabled", onReenabled);
+	
+	// Support for infinite scrolling (only available on Standard not Mobile)
+	const tableBodyNode = document.querySelector("tbody"); // null on mobile
+	const config = {childList: true};
+	const tableBodyMutationObserver = new MutationObserver(onRowsChanged);
+	attachMutationObserverOnStandardSite();
```

Comment:
Please don't attach `MutationObserver`s to the page, it's not a good practice. Please use [`addon.tab.waitForElement`](https://scratchaddons.com/docs/reference/addon-api/addon.tab/#using-addontabwaitforelement) instead.

Reply:
I stumbled across mention of addon.tab.waitForElement as I was figuring things out, but I couldn't find any good documentation about it. The mutation observer, in conjunction with the way Infinite Scrolling works, delivers a single event batching together all the new rows for an entire page at a time. Does waitForElement do anything like that? If not, I'm not exactly sure how to wait for the complete set of rows before I consider all the topic IDs together and redetermine highlights.

If I were to replace the coarse mechanism now, which is basically "reset and recomputer" (to support table manipulation beyond what Infinite Scrolling does) I'd want each topic to determine its relative age based on a window of posts around it _in both directions_, which feels even less tractable using waitForElement, unless there are some associated options or events I haven't seen yet.

I also have in mind another addon, which Necropost Highlighter would need to remain compatible with, that could remove rows from a table. Is there a way waitForElement handles that?

With the infrequency of changes to table rows (only via Infinite Scrolling at this point) is this really an issue? Or is there an extreme cost just to have a MutationObserver attached to a forum page?

Reply:
Hi @Weredime, I think this and someone experienced with regex putting eyes on my URL match in addon.json are the only outstanding items.

I just pushed a change to address the efficiency of updating row changes coming from the Infinite Scrolling addon.

I did a fair bit of research about mutation observers.  (It generally comports with the accepted answer to https://stackoverflow.com/questions/31659567/performance-of-mutationobserver-to-detect-nodes-in-entire-dom which has been kept updated over time and is still often cited). It seems like the kind of Mutation Observer I'm using _should_ be pretty efficient (I'm targeting a single parent, observing infrequent changes to direct children, subtree false, and only observe childList changes.)

However I wasn't treating it like a service interrupt, and the operation I performed in my mutation callback was a bit expensive. Based on a couple sources including the one above, I altered the callback. It now defers the work and immediately return control.

```
  async function onRowsChanged(mutationList, mutationObserver) {
	requestAnimationFrame(rebuild);
  }

  async function rebuild() {
	await removeAnyPriorHighlights();
	highlightNecropostsIfOnSelectedForumPage();
  }
```

Even before the change, the page fetch and parse from Infinite Scrolling dominates everything else so strongly that this work is imperceivable to me.

I profiled in chrome. My computer was fast when I bought it, but that was 12 years ago, so it's pretty low end by today's standards. The animation frame above is taking < 2 ms (as do subsequent layout and paint). There's a layout shift (< 17 ms) associated with Infinite Scrolling that seems unaffected by the presence or absense of Necropost Highlighter.

(I only learned there were profiling tools in a browser today, so hopefully I'm focusing on the right aspects.)

=================================================

I also updated the code that stores original cell clones, in a topic, and restores them later. I now only store/restore the cells that are actually altered. I don't feel a difference when trying it out, but it still feels like a good change.

=================================================

Since Tabs.waitForElement doesn't seem well documented, I read the implementation in Tab.js (and SharedObserver in module.js).

I could rewrite the addon to use waitForElement, but it would be more than a simple drop in replacement for the MutationObserver.

The first design problem I don't have a good solution for is understanding when all rows from the initial page load have been returned by waitForElement. I really can't proceed a row at a time from the beginning, because it won't be until after the rows are available that I can determine what's "old" on the page and begin highlighting.

Unless there's a mechanism I don't see, I'd probably still have to do the initial page load exactly the way I do now. Then I'd have to start a Tabs.waitForElement loop and do bookkeeping to throw away all the rows it returns that match the cells I've already processed from the initial page load.

That would be necessary because the matches have to be in _waitForElementSet later, in order to use markAsSeen to process the post-load changes that Infinite Scrolling cause.

Then, instead of a single batch update of a page-worth of rows (as it is now with my Mutation Observer) it seems the update would look like:

Necropost Highlighter: await addon.tab.waitForElement (in a loop)

Infinite Scrolling: adds a page-worth of rows (let's say 24)

SharedObserver: has a MutationObserver watching the entire document, and will be active, as I'm waiting. For each of the 24 rows, its callback has to do a document.querySelectorAll on my row query and throw away the prior page(s) items as they match the contents of _waitForElementSet, until it finally hits the first new row (match) and resolves it.

Necropost Highlighter: gets the single resolved row, does bookkeeping to make sure it's unseen on this side, and processes the single row. It would be fine to judge the age based on only prior loaded rows at this point, since we have at least a page of prior cellIds. It's not optimal for the interacting with future addons, but it's as good as the current implementation.

This SharedObserver/Necropost Highlighter interplay happens 24 times (with the full document.querySelectorAll in SharedObserver on each of them).

It feels like Tabs.waitForElement and its SharedObserver might be a very valuable optimization for some cases. Especially with many addons in an editor tab, waiting to respond to isolated element changes (and especially to replace many addon observers that all want to service those requests simultaneously).

But Tabs.waitForElement feels out of place for my particular use, processing a batch update on a discussion page.

And as I said up top, the Infinite Scrolling fetch and parse is so time consuming relative to any of this, it's hard to understand the value proposition.

I could well be misunderstanding something, since I'm brand new to javascript. There could be something I'm not seeing and didn't come across in my research.

Is it truly required to rewrite Necropost Highlighter using Tabs.waitForElement, in order to support Infinite Scrolling?

Reply:
I think this is resolved? (since @Weredime suggested a change to how the MutationObserver get constructed, in a review).